Fix building libstd on emscripten targets.

ashleysommer opened the pull request:
The main cause of the problem is that libstd/os/mod.rs treats emscripten targets as an alias of linux targets, whereas liblibc treats emscripten targets as musl-compliant, so it gets a slightly different struct stat64 defined.
This commit adds conditional compilation checks to use the correct timestamp format on fs metadata functions in the case of compiling to emscripten targets.

This commit also depends needs https://github.com/ashleysommer/rust/commit/f1575cff2d631e977038fdba3fa3422ba5f8f2fe applied in order to successfully build libstd with emscripten target.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @brson (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

ashleysommer commented:
This is the error I was getting before the changes:
```
os/linux/fs.rs:19:5: 19:19 error: unresolved import `os::linux::raw`. Could not find `linux` in `os` [E0432]
os/linux/fs.rs:19 use os::linux::raw;
                      ^~~~~~~~~~~~~~
os/linux/fs.rs:19:5: 19:19 help: run `rustc --explain E0432` to see a detailed explanation
os/linux/fs.rs:76:16: 76:65 error: casting `&libc::unix::notbsd::linux::musl::b32::asmjs::stat` as `*const libc::unix::notbsd::linux::musl::b32::asmjs::stat64` is invalid
os/linux/fs.rs:76             &*(self.as_inner().as_inner() as *const libc::stat64
                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
os/linux/fs.rs:105:9: 105:44 error: attempted access of field `st_atime` on type `&libc::unix::notbsd::linux::musl::b32::asmjs::stat`, but no field with that name was found
os/linux/fs.rs:105         self.as_inner().as_inner().st_atime as i64
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
os/linux/fs.rs:108:9: 108:49 error: attempted access of field `st_atime_nsec` on type `&libc::unix::notbsd::linux::musl::b32::asmjs::stat`, but no field with that name was found
os/linux/fs.rs:108         self.as_inner().as_inner().st_atime_nsec as i64
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
os/linux/fs.rs:111:9: 111:44 error: attempted access of field `st_mtime` on type `&libc::unix::notbsd::linux::musl::b32::asmjs::stat`, but no field with that name was found
os/linux/fs.rs:111         self.as_inner().as_inner().st_mtime as i64
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
os/linux/fs.rs:114:9: 114:49 error: attempted access of field `st_mtime_nsec` on type `&libc::unix::notbsd::linux::musl::b32::asmjs::stat`, but no field with that name was found
os/linux/fs.rs:114         self.as_inner().as_inner().st_mtime_nsec as i64
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
os/linux/fs.rs:117:9: 117:44 error: attempted access of field `st_ctime` on type `&libc::unix::notbsd::linux::musl::b32::asmjs::stat`, but no field with that name was found
os/linux/fs.rs:117         self.as_inner().as_inner().st_ctime as i64
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
os/linux/fs.rs:120:9: 120:49 error: attempted access of field `st_ctime_nsec` on type `&libc::unix::notbsd::linux::musl::b32::asmjs::stat`, but no field with that name was found
os/linux/fs.rs:120         self.as_inner().as_inner().st_ctime_nsec as i64
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
sys/unix/fs.rs:153:21: 153:39 error: attempted access of field `st_mtime` on type `libc::unix::notbsd::linux::musl::b32::asmjs::stat`, but no field with that name was found
sys/unix/fs.rs:153             tv_sec: self.stat.st_mtime as libc::time_t,
                                       ^~~~~~~~~~~~~~~~~~
sys/unix/fs.rs:153:31: 153:39 help: did you mean `st_mtim`?
sys/unix/fs.rs:153             tv_sec: self.stat.st_mtime as libc::time_t,
                                                 ^~~~~~~~
sys/unix/fs.rs:154:22: 154:45 error: attempted access of field `st_mtime_nsec` on type `libc::unix::notbsd::linux::musl::b32::asmjs::stat`, but no field with that name was found
sys/unix/fs.rs:154             tv_nsec: self.stat.st_mtime_nsec as libc::c_long,
                                        ^~~~~~~~~~~~~~~~~~~~~~~
sys/unix/fs.rs:154:32: 154:45 help: did you mean `st_mtim`?
sys/unix/fs.rs:154             tv_nsec: self.stat.st_mtime_nsec as libc::c_long,
                                                  ^~~~~~~~~~~~~
sys/unix/fs.rs:160:21: 160:39 error: attempted access of field `st_atime` on type `libc::unix::notbsd::linux::musl::b32::asmjs::stat`, but no field with that name was found
sys/unix/fs.rs:160             tv_sec: self.stat.st_atime as libc::time_t,
                                       ^~~~~~~~~~~~~~~~~~
sys/unix/fs.rs:160:31: 160:39 help: did you mean `st_atim`?
sys/unix/fs.rs:160             tv_sec: self.stat.st_atime as libc::time_t,
                                                 ^~~~~~~~
sys/unix/fs.rs:161:22: 161:45 error: attempted access of field `st_atime_nsec` on type `libc::unix::notbsd::linux::musl::b32::asmjs::stat`, but no field with that name was found
sys/unix/fs.rs:161             tv_nsec: self.stat.st_atime_nsec as libc::c_long,
                                        ^~~~~~~~~~~~~~~~~~~~~~~
sys/unix/fs.rs:161:32: 161:45 help: did you mean `st_atim`?
sys/unix/fs.rs:161             tv_nsec: self.stat.st_atime_nsec as libc::c_long,
                                                  ^~~~~~~~~~~~~
error: aborting due to 11 previous errors
Could not compile `std`.
```

ashleysommer commented:
I'm believe libstd used to build find for emscripten, but I think this commit broke it: https://github.com/rust-lang/rust/commit/aa23c98450063992473d40d707273903f8a3937d

These new fs metadata functions as defined in libstd/os/linux/fs.rs assume the stat64 struct to be defined as per liblibc/src/unix/notbsd/linux/other/b32/mod.rs but instead liblibc uses liblibc/src/unix/notbsd/linux/musl/b32/asmjs.rs for emscripten targets, which has different fields in stat64.

tomaka commented on src/libstd/os/linux/fs.rs:
```diff
@@ -15,6 +15,11 @@ use libc;
+#[allow(deprecated)]
+use os::emscripten::raw;
+
+#[cfg(not(target_os = "emscripten"))]
```
That could be replaced with just `use super::raw;`.

alexcrichton commented:
The `os::linux::fs` module is actually intended to only be compiled on `target_os = "linux"`, could you add a new `os::emscripten` module for the emscripten port? (that was the original intention)

alexcrichton commented on src/libstd/sys/unix/fs.rs:
```diff
@@ -149,32 +155,51 @@ impl FileAttr {
-            tv_sec: self.stat.st_atime as libc::time_t,
-            tv_nsec: self.stat.st_atime_nsec as libc::c_long,
-        }))
+        #[cfg(any(target_env = "musl", target_os = "emscripten"))]
```
Some of these additions of `musl` may not be necessary because the standard library is already guaranteed to compile for `x86_64-unknown-linux-musl`

alexcrichton commented on src/libstd/sys/unix/fs.rs:
```diff
@@ -149,32 +155,51 @@ impl FileAttr {
-            tv_nsec: self.stat.st_birthtime_nsec as libc::c_long,
-        }))
+        #[cfg(any(target_os = "emscripten"))]
+        { return Ok(SystemTime::from( self.stat.st_ctim )); }
```
I think that `ctim` isn't the created time but rather the time of the last status change.

ashleysommer commented:
Ok, thanks for the tips. I reverted all the changes in libstd/os/linux/fs.rs and created a new os submodule called emscripten. I copied over the three module files from os/linux/fs.rs then removed all the unneeded stuff like checking for x86_64 and checking for MIPS.

I also made some changes in the emscripten/linux/raw.rs file so that the `stat` struct better matches the `stat` and `stat64` definition in `liblibc/src/unix/notbsd/linux/musl/b32/asmjs.rs` while maintaining compatibility with the expectations that libstd has about the format of `raw::stat`.

alexcrichton commented on src/libstd/sys/unix/fs.rs:
```diff
@@ -149,22 +155,33 @@ impl FileAttr {
 
     pub fn accessed(&self) -> io::Result<SystemTime> {
-        Ok(SystemTime::from(libc::timespec {
-            tv_sec: self.stat.st_atime as libc::time_t,
```
Could this call `self.st_atime()` and `self.st_atime_nsec()`? I think that would help remove the `#[cfg]`

ashleysommer replied:
I've had a look at doing that, but I don't think it is possible.
`st_atime()` and `st_atime_sec()` are functions available on structs that implement libstd::os::emscripted::fs::MetedataExt or libstd::sys::unix::ext::fs::MetadataExt but I believe the FileAttr struct doesn't implement either of those traits.
In this case the FileAttr struct only has access to its own inner `stat: stat64` struct.

alexcrichton replied:
Oh, right.

In the interest of consistency with other platforms, I think the best solution here would actually be to just modify the libc bindings themselves. Anyone using them manually will require these `#[cfg]` directives anyway, so we may as well get them all aligned.

ashleysommer replied:
Can you elaborate on what you mean here?
Are you saying to change liblibc so that it uses the definitions in `unix/notbsd/linux/other/b32` rather than `unix/notbsd/linux/musl/b32/asmjs` for emscripten targets?
Im not sure if it will be as easy as that, but I can try it out to see how it works.

alexcrichton replied:
Yeah for all other platforms we store the two fields in `st_atim` differently, one for the seconds and one for the nanoseconds. Looks like we accidentally used `timeval` here and we should just switch to two separate fields like in other platofrms.

ashleysommer replied:
I dont think it was an accident. It was most likely copied from here:
https://github.com/kripken/emscripten/blob/incoming/system/lib/libc/musl/arch/emscripten/bits/stat.h

alexcrichton replied:
Yeah, but to mirror how other platforms work and avoid this `#[cfg]` we expand the struct to an equivalent set of individual fields.

ashleysommer replied:
Oh I get it. Yes, I see it the struct will still be byte-compatible and equivalent with individual fields.

alexcrichton replied:
Ah I think it's actually getting taken care of in https://github.com/rust-lang/libc/pull/210, once that's merged this can probably just update libc again

ashleysommer replied:
Ok, I saw the change was merged, so I targeted libc to that commit, and undone these lines.

alexcrichton commented on src/libstd/sys/unix/fs.rs:
```diff
@@ -174,7 +191,8 @@ impl FileAttr {
                   target_os = "freebsd",
-                  target_os = "openbsd")))]
+                  target_os = "openbsd"
+                  )))]
```
This and the above change look like they may no longer be necessary

alexcrichton commented on src/libstd/sys/unix/fs.rs:
```diff
@@ -25,15 +25,21 @@ use sys::time::SystemTime;
 use sys_common::{AsInner, FromInner};
 
-#[cfg(target_os = "linux")]
+#[cfg(any(target_os = "linux", target_env = "musl", target_os = "emscripten"))]
```
Because MUSL implies `target_os = "linux"` I think that clause can be removed from this and the bits below

ashleysommer commented:
Should I squash all these commits together?

ashleysommer commented:
Updated to liblibc e19309c, then took out the extra compile-time cfg checks for emscripten in the FileAttr struct impl. Also rebased to current master, and rebuilt locally and ran all my compilation and usability tests.

alexcrichton commented:
Thanks! Could you also squash the commits down together as well?

alexcrichton commented on src/libstd/sys/unix/fs.rs:
```diff
@@ -156,7 +160,7 @@ impl FileAttr {
 
     pub fn accessed(&self) -> io::Result<SystemTime> {
-        Ok(SystemTime::from(libc::timespec {
+        return Ok(SystemTime::from(libc::timespec {
```
This change can probably be left out for now

ashleysommer replied:
Yep, that was an accident. Shouldn't have been in the commit.

alexcrichton commented:
@bors: r+ 660bbf4f6f4ce7a3d9e4fa7fdfc5e7c87145e049

bors commented:
:hourglass: Testing commit 660bbf4 with merge d5d838b...

bors commented:
:broken_heart: Test failed - [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/3341)

alexcrichton commented:
@bors: retry

On Mon, Mar 7, 2016 at 5:46 AM, bors <notifications@github.com> wrote:

> [image: :broken_heart:] Test failed - auto-win-gnu-64-nopt-t
> <http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/3341>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/31986#issuecomment-193253254>.
>

bors commented:
:hourglass: Testing commit 660bbf4 with merge e06d2ad...

alexcrichton commented:
@bors: retry force

bors commented:
:hourglass: Testing commit 660bbf4 with merge 4352a85...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/8318), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/8343), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/1589), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/8284), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/8288), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/8282), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/1489), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/3442), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/8327), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/8330), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/8298), [auto-mac-ios-opt](http://buildbot.rust-lang.org/builders/auto-mac-ios-opt/builds/457), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/3338), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/3355), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/3348), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/3328), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/2559), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/3336)
